### PR TITLE
kernel: queue: k_queue_poll: Fix slist access race condition

### PR DESCRIPTION
**Pull request by pfalcon**
All sys_slist_*() functions aren't threadsafe and calls to them
must be protected with irq_lock. This is usually done in a wider
caller context, but k_queue_poll() is called with irq_lock already
relinquished, and is thus subject to hard to detect and explain
race conditions, as e.g. was tracked in #4022.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

File: kernel/queue.c [hunk 1]
```diff
@@ -199,6 +199,8 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 {
 	struct k_poll_event event;
 	int err;
+	unsigned int key;
```
Comment (cvinayak) [inline kernel/queue.c]: u32_t ?

Comment (jhedberg) [reply kernel/queue.c]: I think unsigned int is correct. At least that's what irq_lock() is defined to return and irq_unlock() to take as its parameter.

Comment (cvinayak) [reply kernel/queue.c]: sorry... I deleted my comment, asking u32_t. (writing this comment instead again).

**Review by jhedberg — APPROVED** [review]

**Comment by carlescufi** [comment]
@cvinayak @jhedberg and myself are currently testing this patch

**Review by carlescufi — APPROVED** [review]
This fixes some of the issues we have seen this week with the BLE controller communication

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by carlescufi** [comment]
Even when using this fix, we see k_queue_poll() return with NULL. See this issue that is tracking it: https://github.com/zephyrproject-rtos/zephyr/issues/4358

**Review by Vudentz — APPROVED** [review]
Lets have this applied since that is clearly a bug.

**Pull request merged**